### PR TITLE
feat: use own logger, not global logger

### LINKED ISSUE #9: Using `serializable.serializable_class` during import configures root logger
# Description
If some user of py-serializable calls [`serializable.serializable_class`](https://github.com/madpah/serializable/blob/9bc4abccc9cabed5f9808101a8d25717b86f01b4/serializable/__init__.py#L1086) before configuring the root logger (which is common during import blocks), the root logger ends up configured with the default basic configuration.

This happens because we call [`logging.debug` in `ObjectMetadataLibrary.register_klass`](https://github.com/madpah/serializable/blob/9bc4abccc9cabed5f9808101a8d25717b86f01b4/serializable/__init__.py#L969). Calling `logging.debug` (or similar) will [automatically call `logging.basicConfig`](https://docs.python.org/3/library/logging.html#logging.basicConfig) if no handlers are configured for the root logger.

We should probably do what most modules do and set up our own logger and `NullHandler` rather than using the root logger.

# Example
```python
import logging
import serializable

@serializable.serializable_class
class Foo:
    pass

_logger = logging.getLogger()
_logger.addHandler(logging.NullHandler())
_logger.error("foo")
```
Outputs `ERROR:root:foo`, which shouldn't happen if I only configure a `NullHandler`.

# Workaround
I'm currently working around this by wrapping the import of anything that uses `py-serializable` (in this case, CycloneDX) in a logging configuration hack:
```python
_root_logger = logging.getLogger()
_root_null_handler = logging.NullHandler()
_root_logger.addHandler(_root_null_handler)
from cyclonedx.model.bom import Bom
_root_logger.removeHandler(_root_null_handler)
```

# Version Info
- Python 3.10.9 & Python 3.9.13
- py-serializable 0.11.1

### PR DESCRIPTION
**Pull request by kroeschl**
#### Description
Instead of calling `logging.debug()` (or similar) directly and causing a call to `logging.basicConfig()`, log via our logger only. Also stop manipulating the logging level for our module and let users manage that. Fixes #9.

#### Testing
`poetry run tox` succeeds. Also ran the example from the issue and saw no output:
```python
import logging
import serializable

@serializable.serializable_class
class Foo:
    pass

_logger = logging.getLogger()
_logger.addHandler(logging.NullHandler())
_logger.error("foo")
```

### CODE REVIEW

**Comment by jkowalleck** [comment]
some tests are failing. could you get back when they pass?

**Comment by jkowalleck** [comment]
@kroeschl i think all is settled,
could you do a review and see if the feature is still as you expected it?

**Comment by kroeschl** [comment]
> @kroeschl i think all is settled, could you do a review and see if the feature is still as you expected it?

This PR is now a bit more than what I intended to change. Specifically:
- Converting existing warnings to log messages is IMO a design decision outside the scope of this change.
- Using a logger name of `serializable._logging.LOGGER` is counter to [Python logging convention](https://docs.python.org/3/library/logging.html#logger-objects) and may confuse users. Also, a logger name including variations of "logging" is somewhat redundant.
- Generally, adding a logging module is probably overkill since we're not doing anything special in it (aside from the warning `stacklevel` change). Users can already configure logging in their own code by using `getLogger("serializable")` (or whatever name).
- We shouldn't set a default logging level. That should be left up to the user.

I'm inclined to roll this back to just the fix for the issue and let other rework live in a separate PR.

**Comment by jkowalleck** [comment]
thanks for looking into this.


>     * Converting existing warnings to log messages is IMO a design decision outside the scope of this change.

From a architectural perspective, it looked obvious to move all log/output to an own logger.
Warnings were sent to global warning handler, right? So better have them to a logger that you can control downstream.
Would you agree?


>     * Using a logger name of `serializable._logging.LOGGER` is counter to [Python logging convention](https://docs.python.org/3/library/logging.html#logger-objects) and may confuse users. 
Also, a logger name including variations of "logging" is somewhat redundant.

"Python logging convention"? If there is no PEP, then there is no convention, right?
The linked documentation uses just some example to help people chose a unique name for a logger.
And most importantly, accessing the logger is possible via public API `from serializable import LOGGER`, so there is no need to know the name in the first place, or is there?
What do I not see? There must be something I missed.


>     * Generally, adding a logging module is probably overkill since we're not doing anything special in it (aside from the warning `stacklevel` change). [...]

It was needed to prevent circular dependencies ;-) 

>     * [...] Users can already configure logging in their own code by using `getLogger("serializable")` (or whatever name).

The mentioned module is nonpublic API. And so the name is nonpublic.
Should the name be dropped, to make this obvious?

>     * We shouldn't set a default logging level. That should be left up to the user.

You see that the logger is a global one, so there are possibly multiple "users". 
Downstream users would need to add logHandlers in the first place, so they would do log level filtering on the handler.
Let's say library A and B both uses `serializable`, and A wants to listen to Errors only, B wants to listen to warnings and above, then you proposed to:
each library sets the log level on the logger. This would end with a situation that the last setter wins.
In contrast, the current solution is that the logger initiates with the lowest level it is used with by this very lib, and log handlers apply filters 
like in the documentation you must have seen: https://github.com/madpah/serializable/blob/1b39831011b018fca1c3bd0454cc7788f168ebdd/docs/logging.rst?plain=1
Is there something wrong with this approach?

File: serializable/__init__.py [hunk 1]
```diff
@@ -40,15 +39,16 @@
...
 
+from ._logging import _LOGGER, _warning_kwargs
 from .formatters import BaseNameFormatter, CurrentFormatter
 from .helpers import BaseHelper
 
 # !! version is managed by semantic_release
 # do not use typing here, or else `semantic_release` might have issues finding the variable
 __version__ = '0.12.0'
 
-logger = logging.getLogger('serializable')
```
Comment (jkowalleck) [inline serializable/__init__.py:50]: bring back `logger` and mark it as depricated.
 even though `logger` was not used for any logging in the past, it still was public API.

Comment (jkowalleck) [reply serializable/__init__.py:50]: or leave as is and consider it a breaking change -- which is okay for a planned v1.0.0

**Comment by jkowalleck** [comment]
outdated, superseded by #47

**Pull request closed without merging**